python/pdo/contract/response.py

```diff
@@ -90,6 +100,216 @@ class ContractResponse(object) :
+        except:
+            try:
+                e = cls.transaction_submission_exceptions_queue.get_nowait()
+            except:
+                e = None
+
+        return e
+
+    # -------------------------------------------------------
+
+    @classmethod
+    def wait_for_commit(cls, commit_id, use_ledger=True, shut_down=False):
+        """ Wait for completion of the commit task specified by id. Return transaction id if ledger is used, else return None. If shut_down is True,
+        send exit singals to replication and transaction threads. The implementation relies on threading.Condition objects"""
+        
+        if use_ledger:
+            cls.txn_completion_condition.acquire()
+            while commit_id not in cls.set_of_completed_txns: 
+                cls.txn_completion_condition.wait(timeout=1.0)
+                e = cls.get_exceptions_from_past_commits()
+                if e is not None: 
+                    raise Exception(str(e))
+                cls.txn_completion_condition.acquire()    
+            cls.txn_completion_condition.release()    
+
+            global transaction_dependencies
+            contract_id = commit_id[0]
+            state_hash = commit_id[1]
+            txn_id = transaction_dependencies.FindDependencyLocally(contract_id, crypto.byte_array_to_base64(state_hash))
+        
+        else: #wait until the replication task for commit_id completes
+
+            Replicator.replication_completion_condition.acquire()
+            while commit_id not in Replicator.set_of_completed_replications: 
+                Replicator.replication_completion_condition.wait(timeout = 1.0)
+                e = cls.get_exceptions_from_past_commits()
+                if e is not None:
+                    raise Exception(str(e))
+                Replicator.replication_completion_condition.acquire()    
+            Replicator.replication_completion_condition.release()    
+
+            txn_id = None
+
+        if shut_down:
+            cls.exit_commit_workers()
+        
+        return txn_id
+
+    # -------------------------------------------------------
+
+    @classmethod
+    def exit_commit_workers(cls):
+        """Shutdown replication exectuor. Send "exit now" message to txn submit thead"""
+
+        # send termination signal to executor thread 
+        for i in range(Replicator.get_max_num_replicator_threads()):
+            Replicator.pending_replications_queue.put(dict({'exit_now': True}))
+        
+        #shutdown replication executor
+        Replicator.replication_executor.shutdown(wait=True)
+        
+        # send termination signal to txn_executor thread
+        try:
+            cls.pending_transactions_queue.put(dict({'exit_now': True}))
+        except :
+            pass # to cover the case when we run without ledgers, in which case there is no txn thread to terminate 
+
+        #shutdown tx executor
+        cls.transaction_executor.shutdown(wait=True) 
+
```

Comment:
how long does it take to shutdown if you have a storage service that simply holds the requests & times out? Does the client have to wait for all threads to terminate? Or should we just explicitly shut them down.

Reply:
Well, given that the shutdown is a one time call after the client is done with all operations, I prefer a clean shutdown (whatever the method is). Note that any bad storage services from the past would have already been isolated and we would not have outstanding requests against them unless the storage service turned bad in the last "timeout" seconds or so. So in most instances, the shutdown will be immediate after serving all commits.